crates/core/src/config.rs

```diff
@@ -42,62 +38,334 @@ const QUALIFIER: &str = "";
+    {
+        Ok(match level.trim() {
+            "off" | "Off" | "OFF" => LevelFilter::Off,
+            "error" | "Error" | "ERROR" => LevelFilter::Error,
+            "warn" | "Warn" | "WARN" => LevelFilter::Warn,
+            "info" | "Info" | "INFO" => LevelFilter::Info,
+            "debug" | "Debug" | "DEBUG" => LevelFilter::Debug,
+            "trace" | "Trace" | "TRACE" => LevelFilter::Trace,
+            s => return Err(serde::de::Error::custom(format!("unknown log level: {s}"))),
+        })
+    }
+
+    pub fn serialize<S>(level: &LevelFilter, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        let level = match level {
+            LevelFilter::Off => "off",
+            LevelFilter::Error => "error",
+            LevelFilter::Warn => "warn",
+            LevelFilter::Info => "info",
+            LevelFilter::Debug => "debug",
+            LevelFilter::Trace => "trace",
+        };
+        serializer.serialize_str(level)
+    }
+
+    pub fn deserialize<'de, D>(deserializer: D) -> Result<LevelFilter, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let level = <&str>::deserialize(deserializer)?;
+        parse_log_level_str::<D>(level)
+    }
+}
+
+mod serde_option_log_level_filter {
+    use serde::{Deserialize, Deserializer, Serializer};
+    use tracing::log::LevelFilter;
+
+    pub fn serialize<S>(level: &Option<LevelFilter>, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        if let Some(level) = level {
+            super::serde_log_level_filter::serialize(level, serializer)
+        } else {
+            serializer.serialize_none()
+        }
+    }
+
+    pub fn deserialize<'de, D>(deserializer: D) -> Result<Option<LevelFilter>, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let level = <Option<&str>>::deserialize(deserializer)?;
+
+        match level {
+            Some(level) => Ok(Some(
+                super::serde_log_level_filter::parse_log_level_str::<D>(level)?,
+            )),
+            None => Ok(None),
+        }
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize)]
 pub struct Config {
+    /// Node operation mode.
+    pub mode: OperationMode,
+
+    #[serde(flatten)]
+    pub gateway: GatewayConfig,
     pub transport_keypair: TransportKeypair,
+    #[serde(with = "serde_log_level_filter")]
     pub log_level: tracing::log::LevelFilter,
+    #[serde(flatten)]
     config_paths: ConfigPaths,
-    local_mode: AtomicBool,
+}
 
-    #[cfg(feature = "websocket")]
-    #[allow(unused)]
-    pub(crate) ws: WebSocketApiConfig,
+impl Config {
+    pub fn transport_keypair(&self) -> &TransportKeypair {
+        &self.transport_keypair
+    }
 }
 
-#[cfg(feature = "websocket")]
-#[derive(Debug, Copy, Clone)]
-pub(crate) struct WebSocketApiConfig {
-    ip: IpAddr,
-    port: u16,
+#[derive(clap::Parser, Debug, Default, Copy, Clone, Serialize, Deserialize)]
+pub struct GatewayArgs {
```

Comment:
why do we have gatewayconfig and gatewayargs

Reply:
`GatewayArgs` is used for merging configs from cli and config file, it accepts optional IP and port, so that we can know if these values are `None`, we know users are not providing these two options from cli, and we can merge the value in the config file. `GatewayConfig` is the final configuration used across the crate.